plugin/docker/volume.go

```diff
@@ -110,57 +110,82 @@ func (v *volume) createContainer(ctx context.Context, cmd []string) (string, err
 	return created.ID, nil
 }
 
-// Runs cmd in a temporary container. If the exit code is 0, then it returns the cmd's output.
-// Otherwise, it wraps the cmd's output in an error object.
-func (v *volume) runInTemporaryContainer(ctx context.Context, cmd []string) ([]byte, error) {
-	// Create a container that mounts a volume and deletes its file. Run rm -rf on it.
-	cid, err := v.createContainer(ctx, cmd)
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		err := v.client.ContainerRemove(ctx, cid, types.ContainerRemoveOptions{})
+type containerDeleter = func()
+type doInContainerFn = func(containerDeleter) error
+
+// Executes the callback while handling start and cleanup of an existing container.
+func (v *volume) doInTemporaryContainer(ctx context.Context, cid string, fn doInContainerFn) error {
```

Comment:
Could `createContainer`'s signature be extended to `(cid, deleteFunc, error)`, and make the post-condition be that it returns a _running_ container? If so, then I don't think we need `doInTemporaryContainer`.

Reply:
Yeah I think so. At one point `doInTemporaryContainer` handled some of the `defer` calls, but that wasn't flexible enough.